File: 0697_Degree_of_an_Array/solution.cpp
/*
697. Degree of an Array

https://leetcode.com/problems/degree-of-an-array/

求”度”相同的最短子数组的长度。
*/
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
  int findShortestSubArray(vector<int> &nums)
  {
    unordered_map<int, vector<int>> indices;
    for (int i = 0; i < nums.size(); ++i)
      indices[nums[i]].push_back(i);
      
    size_t degree = 0;
    for (const auto &p : indices)
      degree = max(degree, p.second.size());
    int ans = nums.size();
    for (const auto &p : indices)
    {
      if (p.second.size() != degree)
        continue;
      ans = min(ans, p.second.back() - p.second.front() + 1);
    }
    return ans;
  }
};

File: 0697_Degree_of_an_Array/solution.java
package cn.huang.leetcode;

import java.util.HashMap;
import java.util.Map;

/*
    697. Degree of an Array
    Given a non-empty array of non-negative integers nums, the degree of this array is defined as the maximum frequency
     of any one of its elements.

    Your task is to find the smallest possible length of a (contiguous) subarray of nums, that has the same degree as nums.
 */
public class LC_0697_DegreeofanArray {
    public int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> start = new HashMap<Integer, Integer>();
        Map<Integer, Integer> end = new HashMap<Integer, Integer>();
        Map<Integer, Integer> count = new HashMap<Integer, Integer>();

        int maxCount = 0;
        for (int i = 0; i < nums.length; i++) {
            if (!count.containsKey(nums[i])) {
                count.put(nums[i], 0);
                start.put(nums[i], i);
            }
            count.put(nums[i], count.get(nums[i]) + 1);
            end.put(nums[i], i);
            maxCount = Math.max(maxCount, count.get(nums[i]));
        }

        int minLength = Integer.MAX_VALUE;
        for (Integer key : count.keySet()) {
            if (count.get(key) == maxCount) {
                minLength = Math.min(minLength, end.get(key) - start.get(key) + 1);
            }
        }

        return minLength;
    }
}
